src/ast_editor_state.ml

```diff
@@ -0,0 +1,104 @@
+open Import
+
+type t =
+  { mutable webview_map : (string, WebView.t, String.comparator_witness) Map.t
+  ; mutable original_mode : bool
+  ; mutable hover_disposable : Disposable.t option
+  ; mutable pp_doc_to_changed_origin_map :
+      (string, bool, String.comparator_witness) Map.t
+  ; mutable origin_to_pp_doc_map :
+      (string, string, String.comparator_witness) Map.t
+  }
+
+let make () =
+  let webview_map = Map.empty (module String) in
+  let original_mode = true in
+  let hover_disposable = None in
+  let pp_doc_to_changed_origin_map = Map.empty (module String) in
+  let origin_to_pp_doc_map = Map.empty (module String) in
+  { webview_map
+  ; original_mode
+  ; hover_disposable
+  ; pp_doc_to_changed_origin_map
+  ; origin_to_pp_doc_map
+  }
+
+let find_original_doc_by_pp_uri ~uri_string map =
+  let r =
+    Map.filteri ~f:(fun ~key:_ ~data -> String.equal uri_string data) map
+    |> Map.to_alist |> List.unzip |> fst
```

Comment:
What about `Map.find_first`?

Note: `find_first` is the name used in the stdlib, it could be different in base.

Reply:
I can't seem to find another way to search keys by data in `Base`. I sould've probably chosen some other structure like bidirectional map, but I'm not sure it's really needed here since it's the only time i search "the other way around".

Reply:
We could do `Map.filter map ~f:predicate_on_value |> Map.keys`?

Reply:
Right, thanks! I didn't process the fact that it operated on values for some reason. It does look cleaner 0b7301348fc95c6823040df24de11c7bd01ff2bb